[dtensor] add op support for view_as_complex and view_as_real

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122569
* #122541

This PR will unblock DTensor computations for [rotary embeddings](https://github.com/meta-llama/llama/blob/main/llama/model.py#L132) used in LLaMa training.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225 @chauhang

tianyu-l committed:
[dtensor] add op support for view_as_complex and view_as_real

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/ops/view_ops.py:
```diff
@@ -207,13 +218,20 @@ def normalize_sizes(sizes: Union[Shape, Tuple[Shape]]) -> Shape:
         return (InputDim(0),)
     else:
-        return (Flatten.new(tuple(InputDim(i) for i in range(ndim))),)
+        if end_dim < 0:
```
can you add some comment on what's happening there?

wanchaol commented on test/distributed/_tensor/test_view_ops.py:
```diff
@@ -460,6 +460,26 @@ def test_view_ops(self):
             ),
         )
 
+        # TODO: Test with complex numbers cannot be run for the following reason:
```
please add a separate test case to mimic rope embedding case instead as discussed offline

wanchaol reviewed (approved):
awesome work! comments inlined

tianyu-l committed:
Update on "[dtensor] add op support for view_as_complex and view_as_real"


This PR will unblock DTensor computations for [rotary embeddings](https://github.com/meta-llama/llama/blob/main/llama/model.py#L132) used in LLaMa training.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for view_as_complex and view_as_real"


This PR will unblock DTensor computations for [rotary embeddings](https://github.com/meta-llama/llama/blob/main/llama/model.py#L132) used in LLaMa training.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for view_as_complex and view_as_real"


This PR will unblock DTensor computations for [rotary embeddings](https://github.com/meta-llama/llama/blob/main/llama/model.py#L132) used in LLaMa training.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225 chauhang

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
